Support MAVLINK_MSG_ID_SET_ATTITUDE_TARGET

lthall opened the pull request:
This PR supports the full Support MAVLINK_MSG_ID_SET_ATTITUDE_TARGET message as it is specified. It now supports
- quaternion attitude
- quaternion attitude + roll and/or pitch and/or yaw rate
- thrust or climb rate (climb rate is not to spec and must be turned off using GUID_OPTIONS)
- roll and/or pitch and/or yaw rate.

We need to look at how we support the old but incorrect behaviour of attitude roll/pitch + rate yaw.

I also need to add the getter message associated with the setter message.

Closes https://github.com/ArduPilot/ardupilot/issues/14610

hendjoshsr71 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -993,10 +974,15 @@ void ModeGuided::angle_control_run()
     // call attitude controller
-    if (guided_angle_state.use_yaw_rate) {
-        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(roll_in, pitch_in, yaw_rate_in);
+    if (is_zero(guided_angle_state.attitude_quat.length())) {
```
Ahh I now see why you constructed a Quat(0,0,0,0). Might be nicer to just pass in attitude_ignore into here instead for those reading later.  

```suggestion
    if (ignore_attitude) {
```

lthall replied:
We don't have this variable.

lthall replied:
We don't have this parameter. What we have now works so I would suggest we move forward with this and improve it in another PR.

lthall replied:
> Yeah ignore_attitude is up in where the type-mask in parsed in GCS_MAVLINK.
I'm ok with not fixing this for now. But lets leave the conversation open so I or someone else sees to come back to this.
I think it is important we check that the user's input quaternion meets our unit length expectation. But that can be done separately.


I think we can only ensure the quaternion is not zero as they are rarely exactly unit length. This is why we renormalise them regularly to ensure they don't move too far from unit length. We currently use the zero length to denote undefined so we have explicitly handled that case making normalisation safe after that.

hendjoshsr71 replied:
Oh certainly. I didn't mean it had to be exactly unit length, but it should be close coming into the system. If is not "near ||q_vec|| = 1" the math could get weird fast as there are assumptions made that quaternions are near unit length.

tridge replied:
let's add an is_zero() on Quaternion object

lthall commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1054,11 +1054,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         }
 
+        Quaternion attitude_quat;
+        if (attitude_ignore) {
+            attitude_quat = Quaternion(0.0, 0.0, 0.0, 0.0);
+        } else {
+            attitude_quat = Quaternion(packet.q[0],packet.q[1],packet.q[2],packet.q[3]);
+        }
```
Yes, I agree. There are a number of issues here to discuss.

I used the 0,0,0,0 because that was something that has been defined in Mavlink here:
https://mavlink.io/en/messages/common.html#ATTITUDE_QUATERNION
"send [0, 0, 0, 0] if field not supported)"

lthall replied:
> Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??

Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

hendjoshsr71 replied:
> > Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??
> 
> Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

Yeah position_target_int is like that with passing so many variables. Possibly, the solution is to create a structure that holds all the needed variables for these functions and pass those into guided. Then we wouldn't need to change the function definitions which every change or additional feature.

lthall replied:
I would like to leave it like this and get it in and work on cleaning things up after.

hendjoshsr71 replied:
That's fine lets leave this open till it is handled by a later PR

lthall replied:
We still need to check for this to ensure we can normalise the quaternion safely.

lthall replied:
It is also worth noting that we often use zero to denote a parameter is not defined.

hendjoshsr71 replied:
I think I handled this with the check for unit length input

lthall commented on libraries/AP_HAL_SITL/system.cpp:
```diff
@@ -25,7 +25,7 @@ void init()
 }
 
-void WEAK panic(const char *errormsg, ...)
+void panic(const char *errormsg, ...)
```
This was actually a bug in master for Cygwin that I needed to patch here so I could compile. This will be fixed in master and will be able to be removed from here.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:
```diff
@@ -210,40 +210,44 @@ void AC_AttitudeControl::reset_rate_controller_I_terms_smoothly()
-
-    // calculate the attitude target euler angles
-    _attitude_target.to_euler(_euler_angle_target.x, _euler_angle_target.y, _euler_angle_target.z);
+//#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Yeh :)

shubham-shahh commented:
Hi, I am using this PR with [Mavros_controllers](https://github.com/Jaeyoung-Lim/mavros_controllers), the UAV does random motion and then crashes. I would like to know if there's a difference in co-ordinate frames?

lthall commented:
After talking to various Dev members I think we need to:

- Update guided logging to make logging clearer and maybe separate logging for position vs angle commands.
- Add and fix mavlink messages reporting the current state (for example current attitude target rather than the current attitude)
- Add messages to support angle + angle rate + climb rate (maybe)
- Add message to support  angle + angle rate limit + climb rate and/or throttle (maybe).

amilcarlucas commented on ArduCopter/version.h:
```diff
@@ -6,7 +6,7 @@
 #include "ap_version.h"
 
-#define THISFIRMWARE "ArduCopter V4.2.0-dev"
+#define THISFIRMWARE "ArduCopter V4.2.0-GT"
```
This is probably an unwanted change

lthall replied:
Removed

amilcarlucas commented:
Unfortunately this has merge conflicts now.

lthall commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1054,11 +1054,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
             break;
         }
 
+        bool roll_rate_ignore   = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE;
+        bool pitch_rate_ignore  = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
+        bool yaw_rate_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_YAW_RATE_IGNORE;
+        bool throttle_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_THROTTLE_IGNORE;
+        bool attitude_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE;
```
Done

lthall commented:
I have rebased and I think I have addressed most of the problems. Given how much others are doing to further this I would like to get this done so I can get out of the way of others.

hendjoshsr71 commented on libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:
```diff
@@ -210,40 +210,39 @@ void AC_AttitudeControl::reset_rate_controller_I_terms_smoothly()
+
+    // rotate target and normalize
+    Quaternion attitude_desired_update;
+    attitude_desired_update.from_axis_angle(Vector3f{ang_vel_target.x * _dt, ang_vel_target.y * _dt, ang_vel_target.z * _dt});
```
```suggestion
    attitude_desired_update.from_axis_angle(ang_vel_target * _dt);
```

lthall replied:
Yep, nice.

hendjoshsr71 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.cpp:
```diff
@@ -145,16 +145,16 @@ void AC_AttitudeControl_Multi_6DoF::input_angle_step_bf_roll_pitch_yaw(float rol
-// not used anywhere in current code, panic in SITL so this implementaiton is not overlooked
-void AC_AttitudeControl_Multi_6DoF::input_quaternion(Quaternion attitude_desired_quat) {
+// not used anywhere in current code, panic in SITL so this implementation is not overlooked
+void AC_AttitudeControl_Multi_6DoF::input_quaternion(Quaternion& attitude_desired_quat, Vector3f ang_vel_target) {
```
```suggestion
void AC_AttitudeControl_Multi_6DoF::input_quaternion(const Quaternion& attitude_desired_quat, const Vector3f &ang_vel_target) {
```

hendjoshsr71 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -20,7 +20,7 @@ class AC_AttitudeControl_Multi_6DoF : public AC_AttitudeControl_Multi {
     // Command a Quaternion attitude with feedforward and smoothing
     // not used anywhere in current code, panic so this implementaiton is not overlooked
-    void input_quaternion(Quaternion attitude_desired_quat) override;
+    void input_quaternion(Quaternion& attitude_desired_quat, Vector3f ang_vel_target) override;
```
```suggestion
    void input_quaternion(const Quaternion& attitude_desired_quat, const Vector3f &ang_vel_target) override;
```

lthall replied:
I can do this for ang_vel_target but attitude_desired_quat isn't constant.

lthall replied:
opps, no I can't. It is changed when limited.

hendjoshsr71 commented on libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:
```diff
@@ -210,40 +210,39 @@ void AC_AttitudeControl::reset_rate_controller_I_terms_smoothly()
 
 // Command a Quaternion attitude with feedforward and smoothing
-void AC_AttitudeControl::input_quaternion(Quaternion attitude_desired_quat)
+void AC_AttitudeControl::input_quaternion(Quaternion& attitude_desired_quat, Vector3f ang_vel_target)
 {
```
```suggestion
void AC_AttitudeControl::input_quaternion(const Quaternion& attitude_desired_quaternion, const Vector3f &angular_vel_target)
{
    // Inputs
    Quaternion attitude_desired_quat = attitude_desired_quaternion;
    Vector3f ang_vel_target = angular_vel_target;
```

lthall replied:
You are removing an important task of this function.

hendjoshsr71 reviewed (changes requested):
None of these are have to but... I do think we should immediately follow up with checking the user's quaternion to be unit length. 

I haven't yet tested this again since your latest update, but here is a bit of an overview.

lthall commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1093,11 +1093,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
+        const bool throttle_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_THROTTLE_IGNORE;
+        const bool attitude_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE;
+
         // ensure type_mask specifies to use attitude and thrust
```
changed to:
// ensure thrust field is not ignored

lthall commented:
> Yeah ignore_attitude is up in where the type-mask in parsed in GCS_MAVLINK.
I'm ok with not fixing this for now. But lets leave the conversation open so I or someone else sees to come back to this.
I think it is important we check that the user's input quaternion meets our unit length expectation. But that can be done separately.


I think we can only ensure the quaternion is not zero as they are rarely exactly unit length. This is why we renormalise them regularly to ensure they don't move too far from unit length. We currently use the zero length to denote undefined so we have explicitly handled that case making normalisation safe after that.

peterbarker commented on ArduCopter/Copter.cpp:
```diff
@@ -401,7 +401,7 @@ bool Copter::set_target_angle_and_climbrate(float roll_deg, float pitch_deg, flo
     q.from_euler(radians(roll_deg),radians(pitch_deg),radians(yaw_deg));
 
-    mode_guided.set_angle(q, climb_rate_ms*100, use_yaw_rate, radians(yaw_rate_degs), false);
+    mode_guided.set_angle(q, Vector3f(), climb_rate_ms*100, false);
```
```suggestion
    mode_guided.set_angle(q, Vector3f{}, climb_rate_ms*100, false);
```

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1093,11 +1093,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
 
-        // ensure type_mask specifies to use attitude and thrust
-        if ((packet.type_mask & ((1<<7)|(1<<6))) != 0) {
+        const bool roll_rate_ignore   = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE;
```
I wonder if this code would read better if these were inverted so
```suggestion
        const bool use_roll_rate    = !(packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE);
```

... there's a bunch of double-negatives in this function that would disappear further down with this sort of change.

hendjoshsr71 replied:
This is the same pattern we have used right below this in `MAVLINK_MSG_ID_SET_POSITION_TARGET_LOCAL_NED` which  matches to what the mavlink type_mask is which is an ignore_mask not a use_mask.

If we want to change this then lets change it everywhere at once afterwards?

peterbarker commented on ArduCopter/mode_guided.cpp:
```diff
@@ -997,10 +978,15 @@ void ModeGuided::angle_control_run()
     // call attitude controller
-    if (guided_angle_state.use_yaw_rate) {
-        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(roll_in, pitch_in, yaw_rate_in);
+    if (is_zero(guided_angle_state.attitude_quat.length())) {
```
```suggestion
    if (guided_angle_state.attitude_quat.is_zero()) {
```

new method needed

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1093,11 +1093,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
 
+        Quaternion attitude_quat;
+        if (attitude_ignore) {
+            attitude_quat = Quaternion(0.0, 0.0, 0.0, 0.0);
```
```suggestion
            attitude_quat.zero();
```

peterbarker commented on ArduCopter/mode_guided.cpp:
```diff
@@ -617,11 +610,15 @@ void ModeGuided::set_angle(const Quaternion &q, float climb_rate_cms_or_thrust,
     // log target
     copter.Log_Write_GuidedTarget(guided_mode,
-                           Vector3f(guided_angle_state.roll_cd, guided_angle_state.pitch_cd, guided_angle_state.yaw_cd),
+                           Vector3f(ToDeg(roll_rad) * 100.0f, ToDeg(pitch_rad) * 100.0f, ToDeg(pitch_rad) * 100.0f),
```
```suggestion
                           Vector3f{ToDeg(roll_rad), ToDeg(pitch_rad), ToDeg(pitch_rad)} * 100.0f,
```

peterbarker replied:
... similarly elsewhere

hendjoshsr71 reviewed (commented):
I've updated this with the requested changes and done some basic testing using mavproxy's message module.

I've added as a separate commit a check to ensure the input quaternion is near the unit length assumption. 

@lthall you may want to review the changes?

hendjoshsr71 reviewed (approved):
LGTM

hendjoshsr71 commented:
This Closes https://github.com/ArduPilot/ardupilot/issues/14610

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -872,7 +872,7 @@ class ModeGuided : public Mode {
     bool requires_terrain_failsafe() const override { return true; }
 
-    void set_angle(const Quaternion &q, float climb_rate_cms_or_thrust, bool use_yaw_rate, float yaw_rate_rads, bool use_thrust);
+    void set_angle(const Quaternion attitude_quat, const Vector3f &ang_vel, float climb_rate_cms_or_thrust, bool use_thrust);
```
We should have a comment above this method and/or in the .cpp file clarifying that attitude_quat should be zero if only an angular velocity is provided.

Also the "const Quaternion attitude_quat" can be passed by reference.  So it should be, "const Quaternion& attitude_quat"

hendjoshsr71 replied:
I don't think this is quite right. 

This new scheme allows you to specify both the attitude and rates at the same time. If the attitude quaternion is zero then we use only the angular velocity instead through a different function.

I've added a new comment to clear this up hoepfully?

rmackay9 replied:
great, the comment looks very clear now, txs!

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -997,10 +978,15 @@ void ModeGuided::angle_control_run()
-        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(roll_in, pitch_in, yaw_rate_in);
+    if (guided_angle_state.attitude_quat.is_zero()) {
+        attitude_control->input_rate_bf_roll_pitch_yaw(ToDeg(guided_angle_state.ang_vel.x) * 100.0f, ToDeg(guided_angle_state.ang_vel.y) * 100.0f, ToDeg(guided_angle_state.ang_vel.z) * 100.0f);
+    } else if (false) {
```
We don't normally leave commented out code around like this so I think this should be removed.

hendjoshsr71 replied:
Oh thanks! Missed this one!

rmackay9 commented:
Sorry, I'm a little late to the party on this one.  I've found a couple of non-functional issues that could be fixed (see above) although that doesn't need to block the merge of this PR.

I've removed the "MergeOnCIPass" label but that's just so that @hendjoshsr71 sees my comments.  Maybe Josh will want to fix these before he merges them or maybe not.

In any case, I've given this a quick test with the copy-paste example from the MAVLink Interface page on the subject and confirmed that at least the existing functionality still works. https://ardupilot.org/dev/docs/copter-commands-in-guided-mode.html#copter-commands-in-guided-mode-set-attitude-target

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl.h:
```diff
@@ -295,7 +296,7 @@ class AC_AttitudeControl {
     // calculates the velocity correction from an angle error. The angular velocity has acceleration and
     // deceleration limits including basic jerk limiting using smoothing_gain
```
The comment mentions "smoothing_gain" but this argument has been renamed to "input_tc" now.

rmackay9 commented:
This looks good to me.  I'm unsure why it isn't passing the autotests.  I've kicked them off again so if they pass we can merge.  If it doesn't pass then maybe this PR should be rebased on master with the 
hope that perhaps the flapping test has been resolved in master.

rmackay9 commented:
Still not passing so this probably needs a rebase on master.

lthall commented:
That test is the:
  test.CopterTests1e:
    Mount (Test Camera/Antenna Mount) (Mount pitch not achieved) (see /tmp/buildlogs/ArduCopter-Mount.txt)

I was seeing this test fail before too. Does this test consume this message?

hendjoshsr71 commented:
I think i got the auto test failure fixed here. 

The autotest changes here removed the setting of angular rates as they were being used previously before this PR. Making these changes "equivalent" to previous behavior.

This autotest `test.Mount` could probably use some work too, but we shouldn't hold up this PR in my belief.

amilcarlucas commented on Tools/autotest/arducopter.py:
```diff
@@ -4251,9 +4255,9 @@ def test_mount(self):
-                1, # pitch rate
-                1, # yaw rate
+                0, # roll rate (rad/s)
+                0, # pitch rate
+                0, # yaw rate
```
```suggestion
                0, # pitch rate (rad/s)
                0, # yaw rate (rad/s)
```

hendjoshsr71 commented:
Thank you @amilcarlucas for the review and the logging bug fix!

hendjoshsr71 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -617,11 +616,15 @@ void ModeGuided::set_angle(const Quaternion &q, float climb_rate_cms_or_thrust,
     // log target
     copter.Log_Write_GuidedTarget(guided_mode,
-                           Vector3f(guided_angle_state.roll_cd, guided_angle_state.pitch_cd, guided_angle_state.yaw_cd),
+                           Vector3f{ToDeg(roll_rad), ToDeg(pitch_rad), ToDeg(pitch_rad)} * 100.0,
```
Fixed. Wow not how this got missed. Thanks!

hendjoshsr71 commented on ArduCopter/mode.h:
```diff
@@ -872,7 +872,15 @@ class ModeGuided : public Mode {
+    // attitude_quat: IF zero: ang_vel (angular velocity) must be provided even if all zeroes
+    //                IF non-zero: attitude_control is performed using both the attitude quaternion and angular velocity
+    // ang_vel: angular velocity (rad/s)
+    // climb_rate_cms_or_thrust: represents either the climb_rate (m/s) or thrust scaled from [0, 1], unitless
```
Fixed all of these thanks!

IuliuNovac commented:
It looks like there is a bug.

If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.

So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
`    unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
`

shubham-shahh commented:
> It looks like there is a bug.
> 
> If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.
> 
> So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
> `    unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
> `

Are you referring to the issue mentioned here? https://github.com/ArduPilot/ardupilot/issues/20444

IuliuNovac commented:
> > It looks like there is a bug.
> > If you use ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE, then you can use body rates, else, you can't.
> > So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion.However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates.  I am working with 4.4.2, but in master branch the last time i checked it was still the same.
> > `   unsigned int type_mask = ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE | ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;`
> 
> Are you referring to the issue mentioned here? #20444

Thanks, that seems to be right, but holly Christ, there so much emotions in that thread.

lthall commented:
>> So far, based on documentation if you set the mask as below, you should be able to control yaw body rate, while Pitch and Roll from quaternion. However, unless ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE is set, you have no control over yaw rates. I am working with 4.4.2, but in master branch the last time i checked it was still the same.

The problem with this is the quaternion specifies all three axis. There is no way for this message to specify only roll and pitch angles because it's only attitude specification is the single axis quaternion rotation. When we ignore all rates what we are really saying is they are zero. This means the attitude we specify with the quaternion is the target stationary attitude. So when you specify a yaw rate you are really saying roll rate = 0, pitch rate = 0, yaw = something.

The reason for the emotion in the other threads is people want to specify the attitude in Euler angles with Euler rates but this message does not support Euler angles and rates at all. The reaction is to say, no problem I will just convert from Euler to quaternion and ignore the yaw. The problem is this message has no concept of Euler angles and therefore can never be used in this way.

I would like to support an easy Euler based attitude message that does support this input however I suspect that I don't get any real interest in this because everybody discovers they can get all they need and more once they work out how to use the current message correctly. The problem is there are a number of ways to use this message incorrectly that appear to be sensible at first glance.

amilcarlucas commented:
@lthall your videos explain this in detail. It's just a pity that people do not search for the videos. If they would, they would be a lot more productive faster.

IuliuNovac commented:
> @lthall your videos explain this in detail. It's just a pity that people do not search for the videos. If they would, they would be a lot more productive faster.

I am having a hard time finding this videos, could you suggest where to look for them?

lthall commented:
Could you start a thread and repeat this question and the how the set attitude is supposed to be utilized question on the discus forums so we can talk there: https://discuss.ardupilot.org/
